### PR TITLE
Use safe variance epsilon for float16 layer_norm

### PR DESCRIPTION
**Pull request by benbarsdell**
- The original epsilon value (1e-12) is too small for float16 and can
  cause NaNs in the output when the variance is small.
- This commit also adds float16 and float32 cases to LayerNormTest and
  improves the numerical robustness of the test logic.

I believe this is one of (but not all of) the reasons that NaNs are observed during fp16 training of the BERT model as noted here: https://github.com/google-research/bert/pull/255#issuecomment-450591507

### CODE REVIEW

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR!

File: tensorflow/contrib/layers/python/layers/layers.py [hunk 1]
```diff
@@ -2312,7 +2312,9 @@ def layer_norm(inputs,
     norm_axes = list(range(begin_norm_axis, inputs_rank))
     mean, variance = nn.moments(inputs, norm_axes, keep_dims=True)
     # Compute layer normalization using the batch_normalization function.
-    variance_epsilon = 1e-12
+    # Note that epsilon must be increased for float16 due to the limited
+    # representable range.
+    variance_epsilon = 1e-12 if dtype != dtypes.float16 else 1e-3
```
[the remark this answers is not in the document]
Comment (benbarsdell) [reply tensorflow/contrib/layers/python/layers/layers.py]: Choosing the best value here is a bit tricky. Some considerations:

- 1e-3 is the default for all of the TF Python batchnorm APIs (and the FusedBatchNorm op uses 1e-4).
- 1e-7 is in the subnormal range of values for fp16 (not sure if this is a concern or not).
- If the input array is not normally-distributed, the maximum value can far exceed the stddev, which results in overflow during the division if epsilon is too small.
- The BERT model fails to train (gets NaNs) if this variance_epsilon is set to less than ~3e-5.

I'm happy to make the value smaller, but I think 3e-5 is as small as we can practically go (1e-4 seems like a safe choice). Let me know what you think.

[the remark this answers is not in the document]
Comment (tfboyd) [reply tensorflow/contrib/layers/python/layers/layers.py]: @fchollet @benbarsdell ping

**Review by fchollet — APPROVED** [review]
Sounds good, thanks!

**Comment by jlebar** [comment]
> **import/copybara ** — An error happened while migrating the change
Required

@rthadur any idea what this means / how to retrigger?
